gpkit/nomials/map.py

```diff
@@ -0,0 +1,160 @@
+"Implements the NomialMap class"
+from collections import defaultdict
+import numpy as np
+from .. import DimensionalityError
+from ..small_classes import HashVector, Quantity, Strings
+from ..small_scripts import mag
+from ..varkey import VarKey
+from .substitution import parse_subs
+
+
+DIMLESS_QUANTITY = Quantity(1, "dimensionless")
+
+
+class NomialMap(HashVector):
+    "Class for efficent algebraic represention of a nomial"
+    units = None
+    expmap = None
+    csmap = None
+
+    def set_units(self, thing, thing2=None):
+        "Sets units to those of `thing*thing2`"
+        if thing is None and thing2 is None:
+            self.units = None
+        elif hasattr(thing, "units"):
+            if hasattr(thing2, "units"):
+                self.units = Quantity(1, (thing*thing2).units)
+                try:  # faster than "if self.units.dimensionless"
+                    conversion = float(self.units)
+                    self.units = None
+                    for key, value in self.items():
+                        self[key] = value*conversion
+                except DimensionalityError:
+                    pass
+            else:
+                self.units = Quantity(1, thing.units)
+        elif hasattr(thing2, "units"):
+            self.units = Quantity(1, thing2.units)
+        else:
+            self.units = None
+
+    def to(self, units):
+        "Converts a NomialMap to the given units"
+        sunits = self.units if self.units else DIMLESS_QUANTITY
+        nm = self * sunits.to(units).magnitude
+        nm.set_units(units)  # pylint: disable=no-member
+        return nm
+
+    def __add__(self, other):
+        "Adds NomialMaps together"
+        if self.units != other.units:
+            other = float(other.units/self.units)*other
+        hmap = HashVector.__add__(self, other)
+        hmap.units = self.units
+        hmap.remove_zeros(just_monomials=True)   # pylint: disable=no-member, protected-access
+        return hmap
+
+    def remove_zeros(self, just_monomials=False):
+        "Removes zeroed exponents and monomials."
+        # TODO: do this automatically during HashVector operations?
+        posynomial = (len(self) > 1)
+        for key in self.keys():
+            value = self[key]
+            if posynomial and value == 0:
+                del self[key]
+            elif not just_monomials:
+                zeroes = set(vk for vk, exp in key.items() if exp == 0)
+                if zeroes:
+                    del self[key]
+                    for vk in zeroes:
+                        del key[vk]
+                    key._hashvalue = None  # reset hash # pylint: disable=protected-access
+                    self[key] = value + self.get(key, 0)
```

Comment:
Yeah! That would definitely make this more readable, though I don't think it'll simplify in the sense of removing lines of code.